### ISSUE TITLE
On the endianness of TensorProto::tensor_content

### ISSUE THREAD

**Issue by jerry73204** [tensorflow/tensorflow#40436]
This is a re-post from [tensorflow forums](https://groups.google.com/a/tensorflow.org/forum/?utm_medium=email&utm_source=footer#!topic/discuss/6ZaRNUF_--g).

I looked into the implementation of conversion from/to the protobuf type TensorProto, and feel confused on the endianness of the underlying storage.

In my understanding of the decoding process, it checks if TensorProto::tensor_content is empty ([code](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.cc#L928-L931)). If not, it calls the template function Helper::Decode ([code](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.cc#L169-L183)), and calls `base()` to cast the bytes data to desired type ([code](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.cc#L176)).

I searched around the definition of `base()` and found that it simply calls `reinterpret_cast` ([code](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.h#L86-L89)). The endianness of `reinterpret_cast` depends on the system architecture. So does tensorflow goes as it is, or did I miss anything?

**TL;DR** What is the endianness of the protobuf type `TensorProto::tensor_content`?

**Comment by mihaimaruseac** [comment]
You shouldn't care about the endianness of a proto, the library [takes care of it](https://stackoverflow.com/questions/28905274/endianness-of-protocol-buffer-message)

**Comment by jerry73204** [comment]
I agree. I think it is not the case because the tensor data is serialized *before* protobuf serialization if I understand it correctly.

To be precise, `Tensor::tensor_content` is defined as repeated uint8 in protobuf ([code](https://github.com/tensorflow/tensorflow/blob/a975ef0f5c81d913d7bb1a3a2a2e84fe1f255a71/tensorflow/core/framework/tensor.proto#L38)). By putting the tensor buffer, which is contiguous floats or integers, it's reinterpreted into contiguous uint8 (in C++) in anticipation. That's the place the endianness could go wrong.

**Comment by mihaimaruseac** [comment]
If the tensor contains an array of floats then the repeated `float_val` (https://github.com/tensorflow/tensorflow/blob/a975ef0f5c81d913d7bb1a3a2a2e84fe1f255a71/tensorflow/core/framework/tensor.proto#L50) gets used, not `tensor_content`

**Comment by jerry73204** [comment]
@mihaimaruseac yes, and it's partially true. The [decoding code](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.cc#L928-L931) shows it decides on whether `tensor_content` is empty or not to decode `tensor_content` or `*_val` fields. On the other hand, you can grep over the repo and find many places encode the tensor by [AsProtoTensorContent](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.cc#L960-L968), and it has the chance to encode and fill the float/integer/etc buffer into `tensor_content`. (It's called `Variant` in the [code](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/core/framework/tensor.cc#L270-L274)).

So, I think there are chances that our issue occurs. The distinction of `tensor_content` and `*_val` may depend on versions or other factors.

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
